sbol_utilities/sbol3_sbol2_conversion.py

```diff
@@ -404,27 +436,56 @@ def visit_component_definition(self, cd2: sbol2.ComponentDefinition):
                     sbol2.BIOPAX_PROTEIN: sbol3.SBO_PROTEIN,
                     sbol2.BIOPAX_SMALL_MOLECULE: sbol3.SBO_SIMPLE_CHEMICAL,
                     sbol2.BIOPAX_COMPLEX: sbol3.SBO_NON_COVALENT_COMPLEX}
-        types3 = [type_map.get(t, t) for t in cd2.types]
+        types3 = [type_map.get(t, t) for t in comp_def2.types]
         # Make the Component object and add it to the document
-        cp3 = sbol3.Component(cd2.identity, types3, namespace=self._sbol3_namespace(cd2),
-                              roles=cd2.roles, sequences=cd2.sequences)
-        self.doc3.add(cp3)
+        comp3 = sbol3.Component(comp_def2.identity, types3, namespace=self._sbol3_namespace(comp_def2),
+                                roles=comp_def2.roles, sequences=comp_def2.sequences)
+        self.doc3.add(comp3)
+
         # Convert the Component properties not covered by the constructor
-        if cd2.components:
-            raise NotImplementedError('Conversion of ComponentDefinition components '
-                                      'from SBOL2 to SBOL3 not yet implemented')
-        if cd2.sequenceAnnotations:
+        identity_mappings = {}
+        if comp_def2.components:
+            for comp2 in comp_def2.components:
+                self.visit_component(comp2, comp3, identity_mappings)
+
+        if comp_def2.sequenceAnnotations:
             raise NotImplementedError('Conversion of ComponentDefinition sequenceAnnotations '
                                       'from SBOL2 to SBOL3 not yet implemented')
-        if cd2.sequenceConstraints:
+        if comp_def2.sequenceConstraints:
             raise NotImplementedError('Conversion of ComponentDefinition sequenceConstraints '
                                       'from SBOL2 to SBOL3 not yet implemented')
         # Map over all other TopLevel properties and extensions not covered by the constructor
-        self._convert_toplevel(cd2, cp3)
+        self._convert_toplevel(comp_def2, comp3)
+        self.handle_subcomponent_identity_triple_surgery(identity_mappings)
 
-    def visit_component(self, a: sbol2.Component):
+    def visit_component(self, comp2: sbol2.Component, comp3: sbol3.Component, identity_mappings):
         # Priority: 2
-        raise NotImplementedError('Conversion of Component from SBOL2 to SBOL3 not yet implemented')
+        sub3 = sbol3.SubComponent(comp2.identity)
+        sub3.roles = comp2.roles
+        if comp2.roleIntegration:
+            sub3.role_integration = comp2.roleIntegration
+        if comp2.sourceLocations:
+            sub3.source_locations = comp2.sourceLocations
+        sub3.instance_of = comp2.definition
+        comp3.features += [sub3]
+        identity_mappings[sub3.identity] = comp2.identity
+
+    def handle_subcomponent_identity_triple_surgery(self, identity_mappings):
+        with tempfile.TemporaryDirectory() as tmpdir:
```

Comment:
I don't think that you should be reading and writing tempfiles here. This could result in large numbers of complete read-write cycles in the course of a single conversion and be extremely slow.

I would suggesting saving up all of the identity surgery plans into a single shared datastructure, then doing the surgery on the graph like it's done over here: https://github.com/SynBioDex/SBOL-utilities/blob/9331964d61ff7f85ed91eee8e5f96cafb59665c5/sbol_utilities/conversion.py#L27